core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.READY;
+
+import io.grpc.ConnectivityState;
+import io.grpc.Internal;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A forwarding load balancer and holder of currentLb and pendingLb. The pendingLb's helper will not
+ * update balancing state until a subchannel managed by the pendingLB is READY, whence the pendingLb
+ * becomes current.
+ */
+@Internal
+@NotThreadSafe // Must be accessed in SynchronizationContext
+public final class GracefulSwitchLoadBalancer extends ForwardingLoadBalancer {
+
+  @Nullable // never null after init()
+  private LoadBalancer delegate;
+  @Nullable // never null after init()
+  private LoadBalancer currentLb;
+  @Nullable // set to null once the new lb becomes current
+  private LoadBalancer pendingLb;
+
+  private GracefulSwitchLoadBalancer() {}
+
+  /** Initializes an instance. */
+  public static GracefulSwitchLoadBalancer init(LoadBalancerProvider lbProvider, Helper helper) {
```

Comment:
Alternatively, you can assign `delegate` and `currentLb` to no-op LoadBalancers initially, and let the user to call `switchTo()` even for the first LB. That will be a smaller API surface and won't be harder to use.

Reply:
Good idea!